File: training/python/canastra_train/elo.py
"""ELO rating tracker for the GA.

Replaces per-generation score differentials with a persistent rating that
carries across generations. Children inherit their parent's rating, so a
small mutation starts with a meaningful prior rather than from scratch.

Standard ELO: expected = 1 / (1 + 10^((opp - self) / 400)), and
rating += K * (result - expected). Results are 1.0 (win), 0.0 (loss), 0.5
(draw). Updates are in-place and processed in deterministic game order
(fixed by ``batch_layout``), so a resumed run reproduces ratings exactly.
"""

from __future__ import annotations

import numpy as np


class EloTracker:
    """Persistent ELO ratings for a roster (population + hall of fame)."""

    def __init__(
        self,
        size: int,
        k_factor: float = 32.0,
        base: float = 1200.0,
    ) -> None:
        self.ratings = np.full(size, base, dtype=np.float64)
        self.k_factor = k_factor
        self.base = base

    def expected(self, a: int, b: int) -> float:
        return float(1.0 / (1.0 + 10.0 ** ((self.ratings[b] - self.ratings[a]) / 400.0)))

    def update(self, a: int, b: int, result: float) -> None:
        exp_a = self.expected(a, b)
        self.ratings[a] += self.k_factor * (result - exp_a)
        self.ratings[b] += self.k_factor * (1.0 - result - (1.0 - exp_a))

    def batch_update(self, results: list[tuple[int, int, float]]) -> None:
        for a, b, result in results:
            self.update(a, b, result)

    def grow(self, n_new: int, parent_indices: list[int] | None = None) -> None:
        """Extend the ratings array. If ``parent_indices`` is given, each new
        entry inherits the corresponding parent's rating (for HOF champions
        that are children of a known parent); otherwise starts at ``base``."""
        old = self.ratings
        if parent_indices is not None:
            inherited = np.array(
                [old[p] if p < len(old) else self.base for p in parent_indices],
                dtype=np.float64,
            )
            self.ratings = np.concatenate([old, inherited])
        else:
            self.ratings = np.concatenate([old, np.full(n_new, self.base, dtype=np.float64)])

    def __len__(self) -> int:
        return len(self.ratings)

    def copy(self) -> EloTracker:
        tracker = EloTracker(len(self), self.k_factor, self.base)
        tracker.ratings = self.ratings.copy()
        return tracker

File: training-rs/src/elo.rs
//! ELO rating tracker for the GA.
//!
//! Standard ELO: expected = 1 / (1 + 10^((opp - self) / 400)), and
//! rating += K * (result - expected). Children inherit their parent's rating.

pub struct EloTracker {
    pub ratings: Vec<f64>,
    k_factor: f64,
    base: f64,
}

impl EloTracker {
    pub fn new(size: usize) -> Self {
        Self::with_params(size, 32.0, 1200.0)
    }

    /// Reconstruct an EloTracker from saved ratings (for --resume).
    pub fn from_ratings(ratings: Vec<f64>) -> Self {
        Self {
            ratings,
            k_factor: 32.0,
            base: 1200.0,
        }
    }

    pub fn with_params(size: usize, k_factor: f64, base: f64) -> Self {
        Self {
            ratings: vec![base; size],
            k_factor,
            base,
        }
    }

    pub fn expected(&self, a: usize, b: usize) -> f64 {
        1.0 / (1.0 + 10.0_f64.powf((self.ratings[b] - self.ratings[a]) / 400.0))
    }

    pub fn update(&mut self, a: usize, b: usize, result: f64) {
        let exp_a = self.expected(a, b);
        self.ratings[a] += self.k_factor * (result - exp_a);
        self.ratings[b] += self.k_factor * (1.0 - result - (1.0 - exp_a));
    }

    pub fn batch_update(&mut self, results: &[(usize, usize, f64)]) {
        for &(a, b, result) in results {
            self.update(a, b, result);
        }
    }

    pub fn grow(&mut self, n_new: usize, parent_indices: Option<&[usize]>) {
        match parent_indices {
            Some(parents) => {
                let inherited: Vec<f64> = parents
                    .iter()
                    .map(|&p| {
                        if p < self.ratings.len() {
                            self.ratings[p]
                        } else {
                            self.base
                        }
                    })
                    .collect();
                self.ratings.extend(inherited);
            }
            None => self.ratings.extend(std::iter::repeat_n(self.base, n_new)),
        }
    }

    pub fn len(&self) -> usize {
        self.ratings.len()
    }

    pub fn is_empty(&self) -> bool {
        self.ratings.is_empty()
    }
}
